File: C++/052_N-Queens II.cpp
class Solution {
public:
    int totalNQueens(int n) {
        if (n == 0 || n == 2 || n == 3)
            return 0;
        if (n == 1)
            return 1;
        this->n = n;
        col = vector<bool>(n, false);
        ltor = vector<bool>(n << 1, false);
        rtol = vector<bool>(n << 1, false);
        board = vector<string>(n, string(n, '.'));
        backtrack(0);
        return num;
    }
private:
    int n, num;
    vector<bool> col, ltor, rtol;
    vector<string> board;
    void backtrack(int x) {
        if (x >= n) {
            num++;
            return;
        }
        for (int i = 0; i < n; i++) {
            if (!col[i] && !ltor[x + i] && !rtol[n - 1 - i + x]) {
                col[i] = ltor[x + i] = rtol[n - 1 - i + x] = true;
                board[x][i] = 'Q';
                backtrack(x + 1);
                board[x][i] = '.';
                col[i] = ltor[x + i] = rtol[n - 1 - i + x] = false;
            }
        }
    }
};

File: Python/052_N-Queens II.py
class Solution:
    # @param {integer} n
    # @return {integer}
    def totalNQueens(self, n):
        if n == 0 or n == 2 or n == 3:
            return 0
        if n == 1:
            return 1
        self.N = n
        self.col = [False] * n
        self.ltor = [False] * (n << 1)
        self.rtol = [False] * (n << 1)
        self.num = 0
        self._backtrack(0)
        return self.num

    def _backtrack(self, x):
        if x >= self.N:
            self.num += 1
            return
        for i in range(0, self.N):
            if not self.col[i] and not self.ltor[x + i] and not self.rtol[self.N - 1 - i + x]:
                self.col[i] = self.ltor[x + i] = self.rtol[self.N - 1 - i + x] = True
                self._backtrack(x + 1)
                self.col[i] = self.ltor[x + i] = self.rtol[self.N - 1 - i + x] = False